// @sudoku/stores/gameDomain.js
import { writable, get } from 'svelte/store';
import { createSudoku, createGame, createSudokuFromJSON, createGameFromJSON } from '../../../domain/index.js';
import { generateSudoku } from '@sudoku/sudoku';
import { decodeSencode } from '@sudoku/sencode';

function createUserGrid() {
  const userGrid = writable([
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
  ]);

  return {
    subscribe: userGrid.subscribe,

    set: (pos, value) => {
      userGrid.update($userGrid => {
        $userGrid[pos.y][pos.x] = value;
        return $userGrid;
      });
    },

    clear: (pos) => {
      userGrid.update($userGrid => {
        $userGrid[pos.y][pos.x] = 0;
        return $userGrid;
      });
    }
  };
}

export const userGrid = createUserGrid();

function createGameStore() {
  const state = writable({
    game: null,
    canUndo: false,
    canRedo: false
  });

  let gameInstance = null;

  function syncToUserGrid() {
    if (gameInstance) {
      const sudoku = gameInstance.getSudoku();
      const gridData = sudoku.getGrid();
      for (let y = 0; y < 9; y++) {
        for (let x = 0; x < 9; x++) {
          userGrid.set({ x, y }, gridData[y][x]);
        }
      }
    }
  }

  function updateState() {
    if (gameInstance) {
      state.set({
        game: gameInstance,
        canUndo: gameInstance.canUndo(),
        canRedo: gameInstance.canRedo()
      });
      syncToUserGrid();
    }
  }

  const store = {
    subscribe: state.subscribe,

    startNew(difficulty) {
      const gridData = generateSudoku(difficulty);
      const sudoku = createSudoku(gridData);
      gameInstance = createGame({ sudoku });
      updateState();
    },

    startCustom(sencode) {
      const gridData = decodeSencode(sencode);
      const sudoku = createSudoku(gridData);
      gameInstance = createGame({ sudoku });
      updateState();
    },

    guess(pos, value) {
      if (!gameInstance) return;
      const { x, y } = pos;
      gameInstance.guess({ row: y, col: x, value });
      updateState();
    },

    undo() {
      if (!gameInstance) return;
      gameInstance.undo();
      updateState();
    },

    redo() {
      if (!gameInstance) return;
      gameInstance.redo();
      updateState();
    },

    toJSON() {
      if (!gameInstance) return null;
      return gameInstance.toJSON();
    },

    loadFromJSON(json) {
      gameInstance = createGameFromJSON(json);
      updateState();
    }
  };

  return store;
}

export const gameDomain = createGameStore();
